packages/eslint-plugin/tests/rules/typedef.test.ts

```diff
@@ -38,6 +38,14 @@ ruleTester.run('typedef', rule, {
         },
       ],
     },
+    {
+      code: `for (const [key, val] of new Map([['key', 1]])) {}`,
```

Comment:
could you please add a test for nested destructuring

```ts
for (const [[[key]]] of [[['key']]]) {}
for (const [[[{ key }]]] of [[[{ key: 'value' }]]]) {}

for (const {p1: {p2: { p3 }}} of {p1: {p2: {p3: 'value'}}}) {}
for (const {p1: {p2: { p3: [key] }}} of {p1: {p2: {p3: ['value']}}}) {}
```

Reply:
Thinking about it - won't the rule break on these cases anyways?

```ts
  const [[[{ key }]]]: [[[{ key: string ]]] = [[[{ key: 'value' }]]]
//         ^^^^^^^    report ❌
//        ^^^^^^^^^   report ❌
//       ^^^^^^^^^^^  report ❌
//      ^^^^^^^^^^^^^ no report ✅ 
```

Reply:
@bradzacher Added additional tests. About the last case, I think the current implementation doesn't handle that, and the proposed changes will not handle/break that behavior.